File: old/timeit2.jl
using Images, FileIO, DelimitedFiles, CSV, DataFrames, Format, VideoIO, Glob
include("helper.jl")
include("kernels.jl")
include("blobs.jl")
include("header.jl")
include("calibration_helper.jl")

function gaussianPyramid_mod(img_gpu, out_gpu, DoG_gpu, DoGo_gpu, conv_gpu, buffer, height, width, imgWidth, octaves; scales = 5, sigma0 = 1.6, k = -1)
	if k == -1
		k = 2^(1 / (scales - 3))
	end
	time_taken = zeros(Float32, 2)
	accumulative_apron::Int8 = 0
	resample_apron::Int8 = 0
	nImages = width ÷ imgWidth
	blobs_input_l = []
	for octave in 1:octaves
		blobs_input_l = []
		for layer in 1:scales
			threads_column = 1024 #32 * 32
			threads_row = (32, 1024 ÷ 32)

			if layer == 1
				sigma = sigma0 * k^(layer - 1)
				apron::Int8 = ceil(Int, 3 * sigma / 2) * 2
			else
				sigma = sigma0 * k^(layer - 2)
				apron = ceil(Int, 3 * sigma / 2) * 2
			end

			while threads_row[2] - 2 * apron <= 0 && threads_row[1] > 4
				threads_row = (threads_row[1] ÷ 2, threads_row[2] * 2)
			end

			if cld(height, prod(threads_column)) >= 1
				blocks_column = makeThisNearlySquare((
					cld(height, threads_column - 2 * apron),
					width,
				))
				blocks_row = makeThisNearlySquare((
					cld(height, threads_row[1]),
					cld(width, threads_row[2] - 2 * apron),
				))

				shmem_column = threads_column * sizeof(Float32)
				shmem_row = threads_row[1] * threads_row[2] * sizeof(Float32)

				time_taken[1] += CUDA.@elapsed buffer .= 0
				if layer == 1
					if octave == 1
						time_taken[1] += CUDA.@elapsed @cuda threads = threads_column blocks = blocks_column shmem = shmem_column maxregs = 64 col_kernel_strips(
							img_gpu,
							conv_gpu[1],
							buffer,
							Int32(width),
							Int16(height),
							Int16(imgWidth),
							Int8(apron),
						)
						CUDA.synchronize()
						error_count = CuArray([0])
						time_taken[1] += CUDA.@elapsed @cuda threads = threads_row blocks = blocks_row shmem = shmem_row maxregs = 32 row_kernel(
							buffer,
							conv_gpu[1],
							out_gpu[1][1],
							Int16(height),
							Int32(width),
							Int16(imgWidth),
							Int8(apron),
							# error_count,
						)
					else
						# take the previous octave's third last output, resample it and take it as input
						apron = 0
						time_taken[1] += CUDA.@elapsed @cuda threads = 1024 blocks = makeThisNearlySquare(((height * width) ÷ (1024), 1)) resample_kernel_2(out_gpu[octave-1][scales-2], out_gpu[octave][1], height * 2, width * 2)
						accumulative_apron = resample_apron
					end
				else
					# take the previous layer's output as input
					time_taken[1] += CUDA.@elapsed @cuda threads = threads_column blocks = blocks_column shmem = shmem_column maxregs = 64 col_kernel_strips(
						out_gpu[octave][layer-1],
						conv_gpu[layer-1],
						buffer,
						Int32(width),
						Int16(height),
						Int16(imgWidth / 2^(octave - 1)),
						Int8(apron),
					)
					CUDA.synchronize()
					error_count = CuArray([0])
					time_taken[1] += CUDA.@elapsed @cuda threads = threads_row blocks = blocks_row shmem = shmem_row maxregs = 32 row_kernel(
						buffer,
						conv_gpu[layer-1],
						out_gpu[octave][layer],
						Int16(height),
						Int32(width),
						Int16(imgWidth / 2^(octave - 1)),
						Int8(apron),
					)
				end
				CUDA.synchronize()
				push!(blobs_input_l, out_gpu[octave][layer])
				if layer == scales
					threads_blobs = (32, 1024 ÷ 32)
					blocks_blobs = (cld(height, threads_blobs[1] - 2), cld(width, threads_blobs[2] - 2))
					shmem_blobs = threads_blobs[1] * threads_blobs[2] * sizeof(Float32) * 4
					DoG_gpu[octave][2] .= 0
					DoG_gpu[octave][1] .= 0
					CUDA.synchronize()
                    time_taken[2] += CUDA.@elapsed DoGo_gpu[octave][1] .= out_gpu[octave][2] .- out_gpu[octave][1]
                    time_taken[2] += CUDA.@elapsed DoGo_gpu[octave][2] .= out_gpu[octave][3] .- out_gpu[octave][2]
                    time_taken[2] += CUDA.@elapsed DoGo_gpu[octave][3] .= out_gpu[octave][4] .- out_gpu[octave][3]
                    time_taken[2] += CUDA.@elapsed DoGo_gpu[octave][4] .= out_gpu[octave][5] .- out_gpu[octave][4]
					CUDA.synchronize()
				end
				accumulative_apron += apron
				if layer == scales - 2
					resample_apron = accumulative_apron ÷ 2
				end
			end
		end
		height = height ÷ 2
		width = width ÷ 2
		time_taken[1] += CUDA.@elapsed buffer = CUDA.zeros(Float32, height, width)
	end
	return time_taken
end


let
    octaves = 5 # not hard coded in the blobs kernel
    layers = 5 # hard coded in the extractBlobXYs function and orientation kernel

    sigma0 = 1.6
    k = 2^(1 / (layers - 3))

    max = 8
    v = []
    path = "assets/videos/cam3"
    f = vcat(glob("*.mp4", path), glob("*.mov", path), glob("*.MP4", path), glob("*.MOV", path))[1]
    for i in 1:2^max
        try
            push!(v, Float32.(Gray.(getFrame(f, i))))
        catch 
            push!(v, Float32.(Gray.(getFrame(f, i-256))))
        end
    end
    for nImages in  2 .^ Vector(1:max)
        # push!(v, Float32.(Gray.(getFrame(f, 339))))
        # push!(v, Float32.(Gray.(getFrame(f, 202))))

        # load the images
        img = []
        imgWidth = 0
        for i in 1:nImages
            img_temp = v[i]	
            # println("image $i: $(size(img_temp))")
            if i == 1
                img = img_temp
                imgWidth = size(img, 2)
            else
                img = cat(img, img_temp, dims = 2)
            end
        end

        height, width = size(img)

        img_gpu, out_gpu, DoG_gpu, conv_gpu, buffer, XY_gpu, DoGo_gpu, DoG_prev_gpu = getGPUElements(img, height, width, layers, octaves, nImages, sigma0, k)

        runs = 100

        # # ============================= Single Gaussian =============================
        # t = 0.0
        # gaussian_index = 1

        # for run in 1:runs
        #     for i in 1:nImages
        #         threads_column = 1024 #32 * 32
        #         threads_row = (32, 1024 ÷ 32)
        #         apron::Int8 = ceil(Int, 3 * sigma0 / 2) * 2
        #         time_taken = 0.0
        #         while threads_row[2] - 2 * apron <= 0 && threads_row[1] > 4
        #             threads_row = (threads_row[1] ÷ 2, threads_row[2] * 2)
        #         end
                

        #         if cld(height, prod(threads_column)) >= 1
        #             blocks_column = makeThisNearlySquare((
        #                 cld(height, threads_column - 2 * apron),
        #                 width,
        #             ))
        #             blocks_row = makeThisNearlySquare((
        #                 cld(height, threads_row[1]),
        #                 cld(width, threads_row[2] - 2 * apron),
        #             ))

        #             shmem_column = threads_column * sizeof(Float32)
        #             shmem_row = threads_row[1] * threads_row[2] * sizeof(Float32)
        #             buffer .= 0
        #             time_taken += CUDA.@elapsed @cuda threads = threads_column blocks = blocks_column shmem = shmem_column maxregs = 64 col_kernel_strips(
        #                 img_gpu,
        #                 conv_gpu[gaussian_index],
        #                 buffer,
        #                 Int32(width),
        #                 Int16(height),
        #                 Int16(imgWidth),
        #                 Int8(apron),
        #             )
        #             CUDA.synchronize()
        #             time_taken += CUDA.@elapsed @cuda threads = threads_row blocks = blocks_row shmem = shmem_row maxregs = 32 row_kernel(
        #                 buffer,
        #                 conv_gpu[gaussian_index],
        #                 out_gpu[1][1],
        #                 Int16(height),
        #                 Int32(width),
        #                 Int16(imgWidth),
        #                 Int8(apron),
        #             )
        #             if run > 1
        #                 t += time_taken
        #             end
        #         end
        #     end
        # end

        # println("Gaussian time: $(t/(nImages * (runs - 1))) seconds per image @ $nImages images at once")
        
        # ============================= Gaussian Pyramid =============================
        
        t = zeros(Float32, 2)
        for run in 1:runs
            time_taken = gaussianPyramid_mod(img_gpu, out_gpu, DoG_gpu, DoGo_gpu, conv_gpu, buffer, height, width, imgWidth, octaves; scales = layers, sigma0 = sigma0)
            if run > 1
                t += time_taken
            end
        end
        println("Pyramid time: [$(t[1]/(nImages * (runs - 1))), $(t[2]/(nImages * (runs - 1)))] seconds per image @ $nImages images at once")
    
        ## ============================= Blobs =============================
        # t = 0
        # for run in 1:runs
		# 	time_taken = 0.0
        #     time_taken, count, orientations, blobs, XY_gpu = getBlobs(img, height, width, imgWidth, octaves, layers, nImages, sigma0, k, 1, time_taken)
        #     if run > 1
        #         t += time_taken
        #     end
        # end
        # println("Blobs time: $(t/(nImages * (runs - 1))) seconds per image @ $nImages images at once")
    end
end

File: old/timeit2.py
import math, time
import cv2

import glob
import os

import ffmpeg
import numpy as np

prints = False

def get_frame(video_path: str, frame_num: int, fps: int = 25) -> np.ndarray:
    """Extract specific frame using FFmpeg's precise seeking"""
    timestamp = frame_num / fps
    
    try:
        # Probe for video stream information
        probe = ffmpeg.probe(video_path)
        video_stream = next(s for s in probe['streams'] if s['codec_type'] == 'video')
        width = int(video_stream['width'])
        height = int(video_stream['height'])
        
        # Read frame as raw RGB bytes
        out, _ = (
            ffmpeg
            .input(video_path, ss=timestamp)
            .output('pipe:', format='rawvideo', pix_fmt='rgb24', vframes=1)
            .run(capture_stdout=True, capture_stderr=True)
        )
    except ffmpeg.Error as e:
        raise RuntimeError(f"FFmpeg error: {e.stderr.decode()}") from e
    
    # Convert to grayscale float32 array
    frame = np.frombuffer(out, np.uint8).reshape([height, width, 3])
    return cv2.cvtColor(frame, cv2.COLOR_RGB2GRAY).astype(np.float32)

sigma0 = 1.6
octaves = 5
layers = 5
k = pow(2, 1/(layers - 3))

gaussians = []
for i in range(1, layers+1):
    sigma = sigma0 * pow(k, i-1)
    apron = math.ceil(3 * sigma/2) * 2
    gaussian = cv2.getGaussianKernel(2*apron+1, sigma).astype(np.float32)
    gaussians.append(gaussian)
    # print(f"layer {i}: apron {2*apron+1}, sigma {sigma}")
    
for images in [pow(2, i) for i in range(6)]:
    path = "assets/videos/cam3"
    video_files = (
        glob.glob(os.path.join(path, "*.mp4")) +
        glob.glob(os.path.join(path, "*.mov")) + 
        glob.glob(os.path.join(path, "*.MP4")) +
        glob.glob(os.path.join(path, "*.MOV"))
    )
    f = video_files[0]

    v = []
    # v.append(get_frame(f, 339))
    # v.append(get_frame(f, 202))
    for i in range(images):
        try:
            v.append(get_frame(f, i))
        except:
            v.append(get_frame(f, i-256))

    # images = len(v)

    gpu_imgs = []
    results = []
    for i in range(images):
        gpu_img = cv2.cuda_GpuMat()
        gpu_img.upload(v[i])
        gpu_img = gpu_img.convertTo(cv2.CV_32F)
        gpu_imgs.append(gpu_img)

    runs = 100
    # ============================= Single Gaussian =============================
    t = 0
    gaussian_index = 0

    for run in range(runs):
        for i in range(images):
            img = v[i]
            # if run == 0:
            #     print(f"image {i}: {img.shape}")
            
            cuda_horizontal = cv2.cuda.createLinearFilter(
                srcType=cv2.CV_32F,
                dstType=cv2.CV_32F,
                kernel=gaussians[gaussian_index].reshape(1, -1),
            )
            cuda_vertical = cv2.cuda.createLinearFilter(
                srcType=cv2.CV_32F,
                dstType=cv2.CV_32F,
                kernel=gaussians[gaussian_index].reshape(-1, 1),
            )
            cv2.cuda.Stream.Null().waitForCompletion()
            start_time = time.time()
            temp = cuda_horizontal.apply(gpu_imgs[i]) 
            cv2.cuda.Stream.Null().waitForCompletion()
            end_time = time.time()
            if run > 0:
                t += (end_time - start_time)

    print(f"Gaussian time: {t/(images*(runs-1)):.6f} seconds per image @ {images} images at once")

    # ============================= Gaussian Pyramid =============================
    t = [0.0] * 2
    resample_buffer = cv2.cuda_GpuMat()
    for run in range(runs):
        for i in range(images):
            outs = []
            DoGs = []
            for octave in range(octaves):
                local_outs = []
                local_DoG = []
                # ===================================== Gaussian =====================================
                for layer in range(layers):
                    if layer == 0:
                        if run == 0 and i == 0 and prints:
                            print(f"\to{octave}l{layer}: new octave, reinitializing local_outs")
                        local_outs = []
                    if (octave == 0) or (octave > 0 and layer > 0):
                        if run == 0 and i == 0 and prints:
                            print(f"o{octave}l{layer}: performing Gaussian convolution", end=" ")
                        cuda_horizontal = cv2.cuda.createLinearFilter(
                            srcType=cv2.CV_32F,
                            dstType=cv2.CV_32F,
                            kernel=gaussians[layer].reshape(1, -1),
                        )
                        cuda_vertical = cv2.cuda.createLinearFilter(
                            srcType=cv2.CV_32F,
                            dstType=cv2.CV_32F,
                            kernel=gaussians[layer].reshape(-1, 1),
                        )
                        cv2.cuda.Stream.Null().waitForCompletion()
                        start_time = time.time()
                        temp = None
                        if layer == 0:
                            if run == 0 and i == 0 and prints:
                                print("on input image")
                            temp = cuda_horizontal.apply(gpu_imgs[i]) 
                        else:
                            if run == 0 and i == 0 and prints:
                                print(f"on previous layer output")
                            temp = cuda_horizontal.apply(local_outs[layer-1])
                        out = cuda_vertical.apply(temp)
                        cv2.cuda.Stream.Null().waitForCompletion()
                        end_time = time.time()
                        local_outs.append(out)
                        if run > 0:
                            t[0] += (end_time - start_time)
                        if layer == layers - 1:
                            # append the whole layer output to outs
                            outs.append(local_outs)
                            # downsample the out[octave][layer-2] to the next octave, take alternate pixels, don't interpolate
                            if run == 0 and i == 0 and prints:
                                print(f"octave is complete. downsampling o{octave}l{layer-2} to next octave (from o{octave}l{layer})")
                            cv2.cuda.Stream.Null().waitForCompletion()
                            start_time = time.time()
                            resample_buffer = cv2.cuda.resize(local_outs[layer-2], (local_outs[layer-2].size()[0] // 2, local_outs[layer-2].size()[1] // 2), interpolation=cv2.INTER_NEAREST)
                            resample_buffer = resample_buffer.convertTo(cv2.CV_32F)
                            cv2.cuda.Stream.Null().waitForCompletion()
                            end_time = time.time()
                            if run > 0:
                                t[0] += (end_time - start_time)
                    else:
                        if run == 0 and i == 0 and prints:
                            print(f"o{octave}l{layer}: replacing output with downsampled image from previous octave ({octave-1})")
                        local_outs.append(resample_buffer)

                # ======================================== DoG =======================================
                for layer in range(layers-1):
                    if layer == 0:
                        if run == 0 and i == 0 and prints:
                            print(f"\to{octave}l{layer}: new octave, reinitializing local_DoG")                
                        local_DoG = []
                    if run == 0 and i == 0 and prints:
                        print(f"o{octave}l{layer}: performing DoG between o{octave}l{layer} and o{octave}l{layer+1}.")
                    cv2.cuda.Stream.Null().waitForCompletion()
                    start_time = time.time()
                    # subtract the previous layer from the current layer, use cuda
                    DoG = (cv2.cuda.subtract(outs[octave][layer], outs[octave][layer+1]))
                    cv2.cuda.Stream.Null().waitForCompletion()
                    end_time = time.time()
                    local_DoG.append(DoG)
                    if run > 0:
                        t[1] += (end_time - start_time)   
                    if layer == layer - 1:
                        DoGs.append(local_DoG)                
    print(f"Pyramid time: [{', '.join(f'{x/(images * (runs-1)):.6f}' for x in t)}] seconds per image @ {images} images at once")